### PR TITLE
Fix Lambda Kinesis event filtering

### PR DESCRIPTION
**Pull request by danielpoonwj**
### Purpose 

Currently Kinesis data event filtering (originally implemented in https://github.com/localstack/localstack/pull/6212) does not work. This is because the data filters are expected to work on parsed JSON, but currently it's a bytes encoded string.

### Approach

https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventfiltering.html#filtering-streams

We don't always have to check for valid JSON, only if the filter pattern contains data properties
```
Incoming data format: Non-JSON
Filter pattern format for data properties: No filter pattern for data properties
Resulting action: Lambda filters (on the other metadata properties only) based on your filter criteria.
```

If data properties are present but the data cannot be parsed as JSON, it will be dropped.
```
Incoming data format: Non-JSON
Filter pattern format for data properties: Valid JSON
Resulting action: Lambda drops the record.
```

### CODE REVIEW

**Comment by danielpoonwj** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by danielpoonwj** [comment]
recheck

**Comment by danielpoonwj** [comment]
Hi @dominikschubert @dfangl, could you help review this fix?

File: localstack/services/awslambda/lambda_utils.py [hunk 1]
```diff
@@ -429,3 +429,12 @@ def get_lambda_event_filters_for_arn(lambda_arn: str, event_arn: str) -> List[Di
     ]
 
     return event_filter_criterias
+
+
+def has_data_filter_criteria(filters: List[FilterCriteria]) -> bool:
+    for filter in filters:
+        for rule in filter["Filters"]:
```
Comment (dominikschubert) [inline localstack/services/awslambda/lambda_utils.py:436]:
```suggestion
        for rule in filter.get("Filters", []):
```

for additional safety

**Review by dominikschubert — CHANGES_REQUESTED** [review]
Thanks for fixing this :+1: 

Since this adds a new feature, could you please add a simple test for this? It's also fine if you have some awscli scripts/commands that you were using when reproducing this issue. I can then create an integration test based on these, so we cover the fix here.

**Comment by danielpoonwj** [comment]
Hi @dominikschubert and @thrau, sorry for the delay. I came up with a simple repo to reproduce the issue as well as to confirm the fix. Instead of using `awscli` to create the resources, I used `serverless` - hope you don't mind.

https://github.com/danielpoonwj/localstack-kinesis-filtering-demo

**Comment by dominikschubert** [comment]
Hi @danielpoonwj!

First, sorry for the delay here. I'd like to help out now finally getting this over the line, but I will need to make some adjustments to the PR here. Could you maybe grant us access to the PR/fork so we can directly contribute?

You can find instructions on how to do this for your fork here: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

The required option should be called `Allow edits from maintainers.`

Other than that there won't be anything else to do on your part :+1:

**Comment by danielpoonwj** [comment]
> Hi @danielpoonwj!
> 
> First, sorry for the delay here. I'd like to help out now finally getting this over the line, but I will need to make some adjustments to the PR here. Could you maybe grant us access to the PR/fork so we can directly contribute?
> 
> You can find instructions on how to do this for your fork here: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests
> 
> The required option should be called `Allow edits from maintainers.`
> 
> Other than that there won't be anything else to do on your part 👍

Screenshot 2023-11-23 at 6 02 25 PM

Done thanks you! Let me know if you're able to access it

**Comment by japuiglaucia** [comment]
Fix Lambda Kinesis event filtering #6826

**Comment by dominikschubert** [comment]
triggered community integration tests against pro manually @ https://github.com/localstack/localstack/actions/runs/7265552739

**Review by dominikschubert — APPROVED** [review]
Tried to reconcile the changes with latest changes from master. Thanks for your contribution and sorry again that this took a while to get in!

**Pull request merged**